File: project_euler/problems_1-100/problems_11-20/p_14.cpp
// Find the longest sequence in the Collatz for any int n[0] < value.

# include <iostream>

int collatz_length_solver(int start) {
    int length = 1;
    unsigned long long n = start;
    while (n > 1) {
        if (n % 2 == 0) {
            n /= 2;
        } else {
            n *= 3;
            n += 1;
        }
        length += 1;
    }
    return length;
}

int main() {
    unsigned long long start_max = 0;
    std::cout << "enter max ";
    std::cin >> start_max;
    unsigned long long longest_num;
    unsigned long long longest = 0;
    for (unsigned long long start = 0; start <= start_max; start++) {
        int len = collatz_length_solver(start);
        if (len > longest) {
            longest = len;
            longest_num = start;
        }
    }
    std::cout << "\n" << longest;
    std::cout << "\n" << longest_num;
}

File: project_euler/problems_1-100/problems_11-20/p_14.py
"""Find the longest sequence in the Collatz for any int n[0] < value.
I wrote a python version to try using built-in @cache memoisation decorator."""

from functools import cache

@cache
def collatz_length_solver(n: int) -> int:
    length = 1
    while n > 1:
        if n % 2 == 0:
            n //= 2
        else:
            n = n * 3 + 1
        length += 1
    return length

def main():
    max = int(input("num: "))
    highest = 0
    for i in range(max):
        len = collatz_length_solver(i)
        if len > highest:
            highest = len
            best_num = i
    print(best_num)

main()
